File: Android/dz7/app/src/main/java/com/example/myapplication/MainActivity.java
package com.example.myapplication;

import android.os.Bundle;
import android.view.View;
import android.widget.Button;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.ContextCompat;

import es.dmoral.toasty.Toasty; // <-- этот импорт должен работать после синхронизации

public class MainActivity extends AppCompatActivity {

    private int sectionIndex = 0;
    private String[] songSections;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        songSections = getResources().getStringArray(R.array.song_sections);
    }

    public void f(View view) {
        if (view instanceof Button) {
            Button button = (Button) view;
            button.setText(R.string.next_section_button);

            if (songSections.length == 0) {
                return;
            }

            String section = songSections[sectionIndex];
            sectionIndex = (sectionIndex + 1) % songSections.length;

            int toastBackground = ContextCompat.getColor(this, R.color.toast_background);
            int toastText = ContextCompat.getColor(this, R.color.toast_text);

            Toasty.custom(
                    this,
                    section,
                    null,
                    toastBackground,
                    toastText,
                    Toasty.LENGTH_LONG,
                    false,
                    true
            ).show();
        }
    }
}

File: Android/dz11/app/src/main/java/com/example/wallpapermanager/MainActivity.kt
package com.example.wallpapermanager

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.widget.Button
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat

class MainActivity : AppCompatActivity() {

    private var startAfterPermission = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        findViewById<Button>(R.id.btnStart).setOnClickListener {
            if (hasNotificationPermission()) {
                startChangerService()
            } else {
                startAfterPermission = true
                requestNotificationPermission()
            }
        }

        findViewById<Button>(R.id.btnStop).setOnClickListener {
            stopService(Intent(this, WallpaperService::class.java))
            Toast.makeText(this, "Service stopped", Toast.LENGTH_SHORT).show()
        }
    }

    private fun startChangerService() {
        val intent = Intent(this, WallpaperService::class.java)
        try {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) startForegroundService(intent)
            else startService(intent)
            Toast.makeText(this, "Service started", Toast.LENGTH_SHORT).show()
        } catch (e: Exception) {
            Toast.makeText(this, "Start failed: ${e.javaClass.simpleName}", Toast.LENGTH_LONG).show()
        }
    }

    private fun hasNotificationPermission(): Boolean {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.TIRAMISU) return true
        return ContextCompat.checkSelfPermission(
            this,
            Manifest.permission.POST_NOTIFICATIONS
        ) == PackageManager.PERMISSION_GRANTED
    }

    private fun requestNotificationPermission() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            ActivityCompat.requestPermissions(
                this,
                arrayOf(Manifest.permission.POST_NOTIFICATIONS),
                1010
            )
        }
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)

        if (requestCode == 1010) {
            val granted = grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED
            if (granted && startAfterPermission) {
                startAfterPermission = false
                startChangerService()
            } else if (!granted) {
                Toast.makeText(this, "Allow notifications to run the service", Toast.LENGTH_LONG).show()
            }
        }
    }
}
